strawberryfields/backends/fockbackend/ops.py

```diff
@@ -318,6 +338,109 @@ def apply_gate_einsum(mat, state, pure, modes, n, trunc):
     return np.einsum(einstring, mat, state, mat.conj())
 
 
+def apply_twomode_gate(mat, state, pure, modes, n, trunc, gate="BSgate"):
+    """Applies a two-mode gate to a state
+
+    Applies the two-mode gate to the state using custom tensor contractions and
+    the numba compiler for faster application.
+
+    Args:
+        mat (ndarray): The BS operator to be applied to the state
+        state (ndarray): The state that the BS is applied to
+        pure (bool): If the state is pure or mixed
+        modes (list[int]): A list of modes to which the BS is applied
+        n (int): The total number of modes
+        trunc (int): The Hilbert space truncation/cutoff
+        gate (str): the gate which should be called (BSgate, S2gate)
+
+    Returns:
+        ndarray: State where the two-mode operation has been applied
+    """
+    if pure:
+        t1 = modes[0]
+        t2 = modes[1]
+
+        # put the ket-values in front to be operated on in the apply function
+        switch_list_1 = np.arange(n)
+        switch_list_2 = np.arange(n)
+        switch_list_1[[0, t1]] = switch_list_1[[t1, 0]]
+        switch_list_2[[1, t2]] = switch_list_2[[t2, 1]]
```

Comment:
Are these transpositions temporary and required due to the differing conventions in the walrus? If so, they can be removed later, so we can add a TODO.

Reply:
I don't think these ones are related to The Walrus

Reply:
They're not related to The Walrus. `switch_list` is there to move the modes on which the gates should be applied to the front indices of `state`, so that `apply_BS_gate` and `apply_S2_gate` can be used (they'll apply the gates to the first two modes in `state`). Then the modes are moved back to their original indices.